### PR TITLE
[Content] Fix art thumb served when adding items to favourites

### LINKED ISSUE #18565: Kodi v19 - Thumbnails/Icons are displayed instead of posters at Favorites and Android TV Tiles
### Describe the bug

Thumbnails or icons are shown instead of posters since Kodi 19 Alpha1/2 (I can't say the exact version)

## Expected Behavior

The behaviour as it was before.

At Favorites:
- Movies are shown with poster
- TV shows are displayed with poster

For Android TV tiles
- Movies are displayed with posters

## Actual Behavior

At Favorites:
- Movies are shown with the thumbnails
- Series are indicated with a star icon

For Android TV tiles
- Movies are also displayed with thumbnails

### To Reproduce
1. Add a movie and a TV show as favourite
2. Show favorites menu
3. Create a playlist and select it as Android TV tile

### Debuglog

not necessary!

### Screenshots 

Android TV Tiles: https://i.imgur.com/0yf9b60.png
Movies @Favorites: https://i.imgur.com/hosslA6.jpg
TV shows @Favorites: https://i.imgur.com/bfPeHjP.jpg

### Your Environment
Used Operating system:
 - [X] Android
 - [ ] iOS
 - [ ] Linux
 - [ ] OSX
 - [ ] Windows
 - [ ] Windows UWP

Operating system version/name:
 - Kodi version: kodi-20201012-ef3ea1b2-master-arm64-v8a

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by enen92**
## Description
This fixes the thumb/icon being selected when adding movies/shows/seasons to favourites introduced when the old global fallback from poster to thumb was removed. Since this is the second time the issue shows, and to reuse the code, I moved the image selection code to a new class in utils.

## Motivation and Context
Fixes https://github.com/xbmc/xbmc/issues/18565

## How Has This Been Tested?
Runtime tested with the addition of different media items (different media type) to favourites. Used vlc to confirm the library served by upnp was still behaving the same.

## Screenshots (if appropriate):

alt text

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

File: xbmc/utils/ContentUtils.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  {
+    return item.GetArt("poster");
+  }
+
+  // seasons
+  else if (item.HasVideoInfoTag() && item.GetVideoInfoTag()->m_type == MediaTypeSeason &&
+           item.HasArt("poster"))
+  {
+    return item.GetArt("poster");
+  }
```
Comment (phunkyfish) [inline xbmc/utils/ContentUtils.cpp:35]: This feels very complex. If there was a `HasPreferredArtMediaType()` function only a single if condition would be required.

**Review by phunkyfish — CHANGES_REQUESTED** [review]

File: xbmc/utils/ContentUtils.cpp (hunk 1 above)
Comment (enen92) [reply xbmc/utils/ContentUtils.cpp:35]: You mean an helper function that would return the preferred art type provided the mediatype?

Comment (phunkyfish) [reply xbmc/utils/ContentUtils.cpp:35]: Like:

```cpp
const std::string ContentUtils::GetPreferredArtImage(const CFileItem& item)
{
  // movies, tvshow or season
  if (item.HasVideoInfoTag() && HasPreferredArtMediaType(item) &&
      item.HasArt("poster"))
  {
    return item.GetArt("poster");
  }

  return item.GetArt("thumb");
}

bool ContentUtils:: HasPreferredArtMediaType(const CFileItem& item)
{
  return item.GetVideoInfoTag()->m_type == MediaTypeMovie ||
             item.GetVideoInfoTag()->m_type == MediaTypeTvShow ||
             item.GetVideoInfoTag()->m_type == MediaTypeSeason;
}

Comment (enen92) [reply xbmc/utils/ContentUtils.cpp:35]: I see your point, however this approach is not extensible as you're setting always the "poster" type to any mediatype that needs special handling. I'll try to work something out on the same line of your approach

Comment (phunkyfish) [reply xbmc/utils/ContentUtils.cpp:35]: 👍

**Comment by phunkyfish** [comment]
Nice, now the supporting methods you have added can be put in an anonymous namespace unless there is value in them being public?

**Review by DaveTBlake — APPROVED** [review]
Thanks for this fix @enen92, approved for v19 beta.

I will just note that the choice of "thumb" or "poster" does connect with the setting of default art types done respectively in  MusicUtils::GetArtTypesToScan() https://github.com/xbmc/xbmc/blob/306145723ecad9b60e81837ad3d681762bdb8384/xbmc/music/MusicUtils.cpp#L349 
and `CVideoThumbloader`  https://github.com/xbmc/xbmc/blob/306145723ecad9b60e81837ad3d681762bdb8384/xbmc/video/VideoThumbLoader.cpp#L239-L246

**Comment by enen92** [comment]
thanks @DaveTBlake I'll add `MediaTypeVideoCollection` to the list too.

I wonder what to do with `MediaTypeMusicVideo` after https://github.com/xbmc/xbmc/pull/18479. Historically the main art type has always been the thumb however since that PR the choice of artwork to display depends on a skin choice and the core is not aware of that. So any item added to the favourites list/xml has to have a default art type and right now it continues to be the thumbnail (even if the user has estuary configured to display posters instead of thumbs).

**Review by phunkyfish — APPROVED** [review]
Nice

**Comment by DaveTBlake** [comment]
> I wonder what to do with MediaTypeMusicVideo after #18479. 

I'm not sure I have any good answers @enen92. I have to wonder if we should have stuck with "thumb" throughout as the default _identifying_ art for an item regardless of what type of item it is - for an album it is the cover, an artist image, a movie poster, TV show has DVD box art (well used to anyway) and episode a screenshot. But do non-tech (or non-English speaking) users know what a thumb/thumbnail is? So I get the change in video to more human relating "poster" but happy to have resisted changing album to "cover" etc. 

Could adding an item to favourites not set all the various art for the item? It would make the list of favourites a mixed list of types of item with mixed types of art, and leave the skin to decide what to display.

**Comment by enen92** [comment]
Unfortunately favourites is too old and was designed as a simple xml file with name, thumb and media url. There's no context to know if it is a movie, a show, music video, wtv, it just parses a line in favourites.xml per entry. IIRC rants to redesign the favourites system pop up now and then, so I believe the issue is more related to the way if was designed than with the artwork. Even sticking with the original "thumbnail" concept, the same issue for music videos would reflect in favourites after that PR.
So I suggest to keep this as is (using thumb for musicvideos in v19) and maybe think about possible improvements for favourites for v20. Maybe storing more artwork in the favourites.xml for entries along with the mediatype (if available) is enough to handle special cases like this one.

**Comment by DaveTBlake** [comment]
>So I suggest to keep this as is (using thumb for musicvideos in v19) and maybe think about possible improvements for favourites for v20

Yes, agreed. Maybe create a road map issue so this doesn't get lost in the melee again.

**Pull request merged**